File: Valid Palindrome - Leetcode 125/Valid Palindrome - Leetcode 125.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        n = len(s)
        L = 0
        R = n - 1

        while L < R:
            if not s[L].isalnum():
                L += 1
                continue

            if not s[R].isalnum():
                R -= 1
                continue

            if s[L].lower() != s[R].lower():
                return False

            L += 1
            R -= 1

        return True

# Time Complexity: O(n)
# Space Complexity: O(1)

File: Valid Palindrome - Leetcode 125/Valid Palindrome - Leetcode 125.cpp
#include <cctype>
#include <string>
using namespace std;

class Solution {
public:
    bool isPalindrome(string s) {
        int L = 0, R = s.size() - 1;

        while (L < R) {
            while (L < R && !isalnum(s[L])) {
                L++;
            }
            while (L < R && !isalnum(s[R])) {
                R--;
            }
            if (L < R && tolower(s[L]) != tolower(s[R])) {
                return false;
            }
            L++;
            R--;
        }
        return true;
    }
};
